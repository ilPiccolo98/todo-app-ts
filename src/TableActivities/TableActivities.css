th,
td {
  border: 1px black solid;
}

.cell-style {
  max-width: 200px;
  word-wrap: break-word;
}
