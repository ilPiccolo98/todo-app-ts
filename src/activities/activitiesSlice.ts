import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { RootState } from "./activitiesStore";
import Activity from "../types/activity/activity";
import axios, { AxiosError } from "axios";

export enum RetrieveState {
  Error,
  Loading,
  Loaded,
}

export enum AddState {
  Adding,
  Added,
  NotAdded,
  Null,
}

export enum UpdateState {
  Updating,
  Updated,
  NotUpdated,
  Null,
}

export enum DeleteState {
  Deleting,
  Deleted,
  NotDeleted,
  Null,
}

type RetrieveActivities = {
  activities: Activity[];
};

type AddActivity = {
  isItAdded: boolean;
};

type DeleteActivity = {
  isItDeleted: boolean;
};

type UpdateActivity = {
  isItUpdated: boolean;
};

export const fetchAllActivities = createAsyncThunk(
  "activities/fetchAllActivities",
  async (_, thunkAPI) => {
    try {
      const response = await axios.get<RetrieveActivities>(
        "/activities/retrieve"
      );
      return response.data.activities;
    } catch (error) {
      const axiosError = error as AxiosError;
      return thunkAPI.rejectWithValue({ errorMessage: axiosError.message });
    }
  }
);

export const addActivity = createAsyncThunk(
  "activities/addActivity",
  async (
    newActivity: { name: string; description: string; status: boolean },
    thunkAPI
  ) => {
    try {
      const response = await axios.post<AddActivity>("/activities/add", {
        ...newActivity,
      });
      return response.data.isItAdded;
    } catch (error) {
      const axiosError = error as AxiosError;
      return thunkAPI.rejectWithValue({ errorMessage: axiosError.message });
    }
  }
);

export const updateActivity = createAsyncThunk(
  "activities/updateActivity",
  async (activityToUpdate: Activity, thunkAPI) => {
    try {
      const response = await axios.post<UpdateActivity>("/activities/update", {
        ...activityToUpdate,
      });
      return response.data.isItUpdated;
    } catch (error) {
      const axiosError = error as AxiosError;
      return thunkAPI.rejectWithValue({ errorMessage: axiosError.message });
    }
  }
);

export const deleteActivity = createAsyncThunk(
  "activities/deleteActivity",
  async (id: number, thunkAPI) => {
    try {
      const response = await axios.delete<DeleteActivity>(
        `/activities/delete/${id}`
      );
      return response.data.isItDeleted;
    } catch (error) {
      const axiosError = error as AxiosError;
      return thunkAPI.rejectWithValue({ errorMessage: axiosError.message });
    }
  }
);

const activitiesSlice = createSlice({
  name: "activities",
  initialState: {
    activities: new Array<Activity>(),
    addingState: AddState.Null,
    updatingState: UpdateState.Null,
    deletingState: DeleteState.Null,
    retrievingState: RetrieveState.Loading,
  },
  reducers: {
    resetAddState: (state) => {
      state.addingState = AddState.Null;
    },
    resetUpdateState: (state) => {
      state.updatingState = UpdateState.Null;
    },
    resetDeleteState: (state) => {
      state.deletingState = DeleteState.Null;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(addActivity.fulfilled, (state, action) => {
      const isItAdded: boolean = action.payload as boolean;
      if (isItAdded) state.addingState = AddState.Added;
      else {
        state.addingState = AddState.NotAdded;
      }
    });
    builder.addCase(addActivity.pending, (state, action) => {
      state.addingState = AddState.Adding;
    });
    builder.addCase(addActivity.rejected, (state, action) => {
      state.addingState = AddState.NotAdded;
    });
    builder.addCase(updateActivity.fulfilled, (state, action) => {
      const isItUpdated: boolean = action.payload as boolean;
      if (isItUpdated) state.updatingState = UpdateState.Updated;
      else {
        state.updatingState = UpdateState.NotUpdated;
      }
    });
    builder.addCase(updateActivity.pending, (state, action) => {
      state.updatingState = UpdateState.Updating;
    });
    builder.addCase(updateActivity.rejected, (state, action) => {
      state.updatingState = UpdateState.NotUpdated;
    });
    builder.addCase(deleteActivity.fulfilled, (state, action) => {
      const isItDeleted: boolean = action.payload as boolean;
      if (isItDeleted) state.deletingState = DeleteState.Deleted;
      else {
        state.deletingState = DeleteState.NotDeleted;
      }
    });
    builder.addCase(deleteActivity.pending, (state, action) => {
      state.deletingState = DeleteState.Deleting;
    });
    builder.addCase(deleteActivity.rejected, (state, action) => {
      state.deletingState = DeleteState.NotDeleted;
    });
    builder.addCase(fetchAllActivities.fulfilled, (state, action) => {
      state.activities = action.payload as Activity[];
      state.retrievingState = RetrieveState.Loaded;
    });
    builder.addCase(fetchAllActivities.pending, (state, action) => {
      state.retrievingState = RetrieveState.Loading;
    });
    builder.addCase(fetchAllActivities.rejected, (state, action) => {
      state.retrievingState = RetrieveState.Error;
    });
  },
});

export const activitiesSelector = (state: RootState) =>
  state.activities.activities;
export const retrievingStateSelector = (state: RootState) =>
  state.activities.retrievingState;
export const addStateSelector = (state: RootState) =>
  state.activities.addingState;
export const updateStateSelector = (state: RootState) =>
  state.activities.updatingState;
export const deleteStateSelector = (state: RootState) =>
  state.activities.deletingState;

export const {
  resetAddState,
  resetDeleteState,
  resetUpdateState,
} = activitiesSlice.actions;

export default activitiesSlice.reducer;
