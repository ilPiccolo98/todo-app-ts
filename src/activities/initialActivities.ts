export type Activity = {
  id: number;
  name: string;
  description: string;
  status: boolean;
};

let idCounter = 1;
export const createActivity = (
  name: string,
  description: string,
  status: boolean
) => {
  const newActivity: Activity = {
    id: idCounter,
    name,
    description,
    status,
  };
  ++idCounter;
  return newActivity;
};

const getInitialActivities = (): Array<Activity> => {
  return [
    createActivity("work out", "after did my homework I must work out", true),
    createActivity("chores", "after the workout I must do the homework", true),
    createActivity(
      "rest",
      "after did everything I deserve a rest thereby to recover",
      false
    ),
  ];
};

export default getInitialActivities;
