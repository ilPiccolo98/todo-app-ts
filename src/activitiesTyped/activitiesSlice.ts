import getInitialState from "./initialActivities";
import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import VectorActivity from "../types/classes/vectorActivity";
import Activity, { ActivityPlain } from "../types/classes/activity";
import { RootState } from "./activitiesStore";

type UpdateActionParameter = {
  id: number;
  name: string;
  description: string;
  status: boolean;
};

type DeleteActionParameter = {
  id: number;
};

type AddActionParameter = {
  name: string;
  description: string;
  status: boolean;
};

const activitiesSlice = createSlice({
  name: "activities",
  initialState: getInitialState(),
  reducers: {
    addActivity: (status, payload: PayloadAction<AddActionParameter>) => {
      const vectorActivity = VectorActivity.fromArrayPlainActivityToVectorActivity(
        status as ActivityPlain[]
      );
      vectorActivity.addActivity(
        new Activity(
          payload.payload.name,
          payload.payload.description,
          payload.payload.status
        )
      );
      status = vectorActivity.toPlainArrayWithPlainActivities();
    },
    deleteActivity: (status, payload: PayloadAction<DeleteActionParameter>) => {
      const vectorActivity = VectorActivity.fromArrayPlainActivityToVectorActivity(
        status as ActivityPlain[]
      );
      vectorActivity.deleteActivity(payload.payload.id);
      status = vectorActivity.toPlainArrayWithPlainActivities();
    },
    updateActivity: (status, payload: PayloadAction<UpdateActionParameter>) => {
      const vectorActivity = VectorActivity.fromArrayPlainActivityToVectorActivity(
        status as ActivityPlain[]
      );
      vectorActivity.updateActivity(
        payload.payload.id,
        payload.payload.name,
        payload.payload.description,
        payload.payload.status
      );
      status = vectorActivity.toPlainArrayWithPlainActivities();
    },
  },
});

export const {
  addActivity,
  deleteActivity,
  updateActivity,
} = activitiesSlice.actions;

export const activitiesSelector = (state: RootState) => state.activities;
export default activitiesSlice.reducer;
