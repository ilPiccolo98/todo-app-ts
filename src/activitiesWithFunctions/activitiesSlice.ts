import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import getInitialActivities from "./initialActivities";
import {
  addActivity as addActivityVector,
  deleteActivity as deleteActivityVector,
  updateActivity as updateActivityVector,
} from "../typesWithoutClasses/vectorActivity/vectorActivity";
import { createActivity } from "../typesWithoutClasses/activity/activity";
import { RootState } from "./activitiesStore";

type AddActivity = {
  name: string;
  description: string;
  status: boolean;
};

type DeleteActivity = {
  id: number;
};

type UpdateActivity = {
  id: number;
  name: string;
  description: string;
  status: boolean;
};

const activitiesSlice = createSlice({
  name: "activities",
  initialState: getInitialActivities(),
  reducers: {
    addActivity: (status, payload: PayloadAction<AddActivity>) => {
      addActivityVector(
        status,
        createActivity(
          payload.payload.name,
          payload.payload.description,
          payload.payload.status
        )
      );
    },
    deleteActivity: (status, payload: PayloadAction<DeleteActivity>) => {
      deleteActivityVector(status, payload.payload.id);
    },
    updateActivity: (status, payload: PayloadAction<UpdateActivity>) => {
      updateActivityVector(
        status,
        payload.payload.id,
        payload.payload.name,
        payload.payload.description,
        payload.payload.status
      );
    },
  },
});

export const {
  addActivity,
  updateActivity,
  deleteActivity,
} = activitiesSlice.actions;

export const activitiesSelector = (state: RootState) => state.activities;

export default activitiesSlice.reducer;
