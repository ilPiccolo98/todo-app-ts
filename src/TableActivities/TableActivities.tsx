import React from "react";
import { activitiesSelector } from "../activities/activitiesSlice";
import { Activity } from "../activities/initialActivities";
import { useSelector } from "react-redux";
import { RootState } from "../activities/activitiesStore";

const TableActivities: React.FC = (): JSX.Element => {
  const activities: Array<Activity> = useSelector<RootState, Array<Activity>>(
    activitiesSelector
  );
  const generateBodyTable = (activities: Array<Activity>): JSX.Element[] => {
    return activities.map(
      (activity: Activity): JSX.Element => {
        return (
          <tr key={activity.id}>
            <td>{activity.id}</td>
            <td>{activity.name}</td>
            <td>{activity.description}</td>
            <td>{activity.status.toString()}</td>
          </tr>
        );
      }
    );
  };

  return (
    <table data-testid="table-activities">
      <tbody>
        <tr>
          <th>Id</th>
          <th>Name</th>
          <th>Description</th>
          <th>Status</th>
        </tr>
        {generateBodyTable(activities)}
      </tbody>
    </table>
  );
};

export default TableActivities;
