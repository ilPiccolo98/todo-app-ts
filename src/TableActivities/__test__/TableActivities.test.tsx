import React from "react";
import { render, RenderResult } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import * as actions from "../../activities/activitiesSlice";
import TableActivities from "../TableActivities";
import { Provider } from "react-redux";
import store from "../../activities/activitiesStore";

const renderTableActivity = () => {
  return render(
    <Provider store={store}>
      <TableActivities />
    </Provider>
  );
};

const getTableBody = (
  component: RenderResult<
    typeof import("@testing-library/dom/types/queries"),
    HTMLElement
  >
): HTMLBodyElement => {
  return component.getByTestId("table-activities")
    .childNodes[0] as HTMLBodyElement;
};

const getCells = (tableBody: HTMLBodyElement, index: number) => {
  const idCell = tableBody.childNodes[index].childNodes[0];
  const nameCell = tableBody.childNodes[index].childNodes[1];
  const descriptionCell = tableBody.childNodes[index].childNodes[2];
  const statusCell = tableBody.childNodes[index].childNodes[3];
  return { idCell, nameCell, descriptionCell, statusCell };
};

describe("testing the TableActivities component", () => {
  const activitiesSelectorSpy = jest.spyOn(actions, "activitiesSelector");
  const retrievingStateSelectorSpy = jest.spyOn(
    actions,
    "retrievingStateSelector"
  );

  beforeEach(() => {
    activitiesSelectorSpy.mockReset().mockReturnValue([]);
    retrievingStateSelectorSpy
      .mockReset()
      .mockReturnValue(actions.RetrieveState.Loaded);
  });

  it("should have 1 activity", () => {
    const activitiesSelectorSpy = jest.spyOn(actions, "activitiesSelector");
    activitiesSelectorSpy.mockReset().mockReturnValue([
      {
        id: 1,
        name: "activity1",
        description: "activity1",
        status: true,
      },
    ]);
    const component = renderTableActivity();
    const tableBody = getTableBody(component);
    const { idCell, nameCell, descriptionCell, statusCell } = getCells(
      tableBody,
      1
    );
    expect(tableBody.childElementCount).toBe(2);
    expect(idCell.textContent).toBe("1");
    expect(nameCell.textContent).toBe("activity1");
    expect(descriptionCell.textContent).toBe("activity1");
    expect(statusCell.textContent).toBe("true");
    expect(activitiesSelectorSpy).toHaveBeenCalled();
    expect(component).toMatchSnapshot();
  });

  it("should show the message: 'Loading...' when retrievingState === RetrieveState.Loading", () => {
    retrievingStateSelectorSpy
      .mockReset()
      .mockReturnValue(actions.RetrieveState.Loading);
    const component = renderTableActivity();
    const message = component.getByText(/Loading.../i);
    expect(message.textContent).toBe("Loading...");
    expect(component).toMatchSnapshot();
  });

  it("should show the message: 'Activities not retrieved!' when retrievingState === RetrieveState.Error", () => {
    retrievingStateSelectorSpy
      .mockReset()
      .mockReturnValue(actions.RetrieveState.Error);
    const component = renderTableActivity();
    const message = component.getByText(/Activities not retrieved!/i);
    expect(message.textContent).toBe("Activities not retrieved!");
    expect(component).toMatchSnapshot();
  });
});
