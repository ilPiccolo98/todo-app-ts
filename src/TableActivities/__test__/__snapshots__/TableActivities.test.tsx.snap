// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the TableActivities component should have 1 activity 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        data-testid="table-activities"
      >
        <tbody>
          <tr>
            <th>
              Id
            </th>
            <th>
              Name
            </th>
            <th>
              Description
            </th>
            <th>
              Status
            </th>
          </tr>
          <tr>
            <td>
              <div
                class="cell-style"
              >
                1
              </div>
            </td>
            <td>
              <div
                class="cell-style"
              >
                activity1
              </div>
            </td>
            <td>
              <div
                class="cell-style"
              >
                activity1
              </div>
            </td>
            <td>
              <div
                class="cell-style"
              >
                true
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      data-testid="table-activities"
    >
      <tbody>
        <tr>
          <th>
            Id
          </th>
          <th>
            Name
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
        </tr>
        <tr>
          <td>
            <div
              class="cell-style"
            >
              1
            </div>
          </td>
          <td>
            <div
              class="cell-style"
            >
              activity1
            </div>
          </td>
          <td>
            <div
              class="cell-style"
            >
              activity1
            </div>
          </td>
          <td>
            <div
              class="cell-style"
            >
              true
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`testing the TableActivities component should show the message: 'Activities not retrieved!' when retrievingState === RetrieveState.Error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="message-error undefined medium"
      >
        <b>
          Activities not retrieved!
        </b>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="message-error undefined medium"
    >
      <b>
        Activities not retrieved!
      </b>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`testing the TableActivities component should show the message: 'Loading...' when retrievingState === RetrieveState.Loading 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="undefined loading-size-medium loading-style"
      >
        Loading...
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="undefined loading-size-medium loading-style"
    >
      Loading...
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
